util/translate_fight.py

```diff
@@ -28,6 +31,145 @@
+            if buff_match:
+                buff_match = buff_match.group(2).split(' from')[0]
+                if not buff_match in ['status']:
+                    translatables['status'].append(buff_match)
+                continue
+
+            # Currently only looks for zone seal messages from log lines.
+            # FIXME: Add other log line possibilities here if necessary.
+            log_match = e_tools.is_tl_line_log(sync_match.group(1))
+            if log_match and 'will be sealed off' in log_match.group(2):
+                if not log_match.group(2).split(' will be')[0] in translatables['PlaceName']:
+                    translatables['PlaceName'].append(log_match.group(2).split(' will be')[0])
+
+            add_match = e_tools.is_tl_line_adds(sync_match.group(1))
+            if add_match:
+                if not add_match.group(1) in translatables['bnpcname']:
+                    translatables['bnpcname'].append(add_match.group(1))
+    return translatables
+
+def compile_and_send(translatables):
+    # XIVAPI/cafemaker allows for returning multiple items in one request if the item IDs are known.
+    IDstr = ",".join(translatables['action_ids'])
+
+    # These columns will be used for every XIVAPI request, so we alias them now.
+    columns = 'Name_en,Name_de,Name_fr,Name_ja'
+    ID_req = requests.get('https://xivapi.com/action', params={'ids': IDstr, 'columns': columns})
+    ID_JSON = ID_req.json()['Results']
+
+    finished_items = defaultdict(dict)
+    for element in ['abilities', 'status', 'bnpcname', 'PlaceName']:
+        finished_items[element] = defaultdict(dict)
+
+    # Nested loops to unroll the abilities into something usable, yay!
+    for ability in range(0, len(ID_JSON)):
+        # Some abilities have no names.
+        if ID_JSON[ability]['Name_en'] is '':
+            continue
+        for lang in xiv_langs:
+            finished_items['abilities'][lang][ID_JSON[ability]['Name_en']] = ID_JSON[ability][xiv_lang_map[lang]]
+
+    # Clearing the actions from the translation dict so we can iterate over it cleanly.
+    translatables.pop('action_ids')
+
+    # Because XIVAPI imposes rate limits, we manually stay under.
+    requestcount = 1
+
+    # Because we don't have the IDs for other elements, we have to search for them individually.
+    # The Requests module blocks by default, so we don't need any asynchronous handling here.
+    # TODO: Look up IDs in St Coinach so we can collapse all this into one request per type?
+    # Nested loops are just THE BEST
+    for element in translatables:
+        working_dict = finished_items[element]
+        if element is 'PlaceName':
+            working_dict = finished_items['bnpcname']
+        for item in translatables[element]:
+            # To avoid being timed out by the server, we manually slow down here.
+            # TODO: Allow for users to include their XIVAPI developer key to increase this limit.
+            if requestcount % 11 is 0:
```

Comment:
Is it possible to make an object that wraps `requests` and does this sleeping automatically?

Reply:
Probably? I've done nothing with Python wrappers before but I can always learn. (Given how things are going though, that might end up being something I add in later due to time constraints.)

Reply:
I meant wrapper in a colloquial sense.  I was thinking that you could make some Python class that either extends `requests` (or has-a `requests`) that you can call get on and hide the `requestcount` inside that class.

Reply:
Mm. Possibly. I'm still not really sure what that would look like, but again I haven't touched that aspect of Python before. Always more to learn!